File: jar/src/main/java/jline/opt/objectives/Objective.java
package jline.opt.objectives;

import jline.opt.results.EvaluationResult;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Abstract base for line-opt objectives. An objective defines the scalar to
 * minimize, optionally with attached constraints folded in as penalties by
 * {@link #evaluateWithPenalty}. Mirrors native-Python
 * {@code line_solver.opt.objectives.Objective}.
 */
public abstract class Objective {

    protected List<Constraint> constraints = new ArrayList<Constraint>();

    public List<Constraint> getConstraints() {
        return constraints;
    }

    /** The scalar to minimize (before constraint penalties). */
    public abstract double evaluate(EvaluationResult result, Map<String, Object> variableValues);

    public abstract boolean isMinimization();

    /** Objective value plus penalty_weight times total constraint violation. */
    public double evaluateWithPenalty(EvaluationResult result,
                                      Map<String, Object> variableValues,
                                      double penaltyWeight) {
        double obj = evaluate(result, variableValues);
        double penalty = 0.0;
        for (Constraint c : constraints) {
            penalty += c.evaluate(result, variableValues) * penaltyWeight;
        }
        return obj + penalty;
    }
}

File: matlab/src/opt/+opt/Objective.m
classdef Objective < handle
    % Objective  Abstract base for line-opt objectives. Mirrors native-Python
    % line_solver.opt.objectives.Objective: defines the scalar to minimize,
    % with attached constraints folded in as penalties by evaluateWithPenalty.

    properties
        constraints = {};   % cell array of opt.Constraint
    end

    methods
        function c = getConstraints(obj)
            c = obj.constraints;
        end
        function v = evaluateWithPenalty(obj, result, variableValues, penaltyWeight)
            if nargin < 4, penaltyWeight = 1e6; end
            v = obj.evaluate(result, variableValues);
            for i = 1:numel(obj.constraints)
                v = v + obj.constraints{i}.evaluate(result, variableValues) * penaltyWeight;
            end
        end
    end

    methods (Abstract)
        v = evaluate(obj, result, variableValues)
        tf = isMinimization(obj)
    end

    methods (Static)
        function d = numericValue(value)
            if isempty(value)
                d = 0.0;
            elseif isnumeric(value)
                d = sum(value(:));
            else
                d = 0.0;
            end
        end
        function tf = isScalarNumeric(value)
            tf = isnumeric(value) && isscalar(value);
        end
    end
end
